File: example3/src/main/java/javatest/colorhaake/com/example3/view/main/MainActivity.java
package javatest.colorhaake.com.example3.view.main;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import javatest.colorhaake.com.example3.R;
import javatest.colorhaake.com.example3.Example3App;
import javatest.colorhaake.com.example3.injection.component.ActivityComponent;
import javatest.colorhaake.com.example3.injection.component.DaggerActivityComponent;
import javatest.colorhaake.com.example3.injection.module.ActivityModule;

import java.util.List;

import javax.inject.Inject;

public class MainActivity extends AppCompatActivity implements MainMvpView {

    public static final String TAG = MainActivity.class.getName();

    @Inject MainPresenter mMainPresenter;

    private ActivityComponent activityComponent;
    public ActivityComponent activityComponent() {
        if (activityComponent == null) {
            activityComponent = DaggerActivityComponent.builder()
                    .activityModule(new ActivityModule(this))
                    .applicationComponent(Example3App.get(this).component())
                    .build();
        }
        return activityComponent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        activityComponent().inject(this);

        mMainPresenter.attachView(this);
        mMainPresenter.viewReady(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        mMainPresenter.detachView();
    }


    /***** MVP View methods implementation *****/

    @Override
    public void showMainPage(List<String> list) { }
}

File: example3/src/main/java/javatest/colorhaake/com/example3/view/main/MainActivity.kt
package javatest.colorhaake.com.example3.view.main

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager

import javatest.colorhaake.com.example3.R
import javatest.colorhaake.com.example3.Example3App
import javatest.colorhaake.com.example3.injection.component.ActivityComponent
import javatest.colorhaake.com.example3.injection.component.DaggerActivityComponent
import javatest.colorhaake.com.example3.injection.module.ActivityModule
import javatest.colorhaake.com.example3.model.ImageData
import javatest.colorhaake.com.example3.model.Response
import kotlinx.android.synthetic.main.activity_main.*

import javax.inject.Inject

class MainActivity : AppCompatActivity(), MainMvpView {

    @Inject
    lateinit var mMainPresenter: MainPresenter

    private val adapter: MainAdapter by lazy {
        MainAdapter()
    }

    private val component: ActivityComponent by lazy {
        DaggerActivityComponent.builder()
                .activityModule(ActivityModule(this))
                .applicationComponent(Example3App[this].component())
                .build()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        component.inject(this)

        recycler_view.adapter = adapter
        recycler_view.layoutManager = LinearLayoutManager(this)

        mMainPresenter.attachView(this)
        mMainPresenter.viewReady()
    }

    private fun onSeachTextChange(term: String) {
        mMainPresenter.searchImages(term, 1)
    }

    override fun onStop() {
        super.onStop()
        mMainPresenter.detachView()
    }


    /***** MVP View methods implementation  */

    override fun showMainPage(searchImageData: Response<ImageData>) {
        adapter.updateSearchImageData(searchImageData, ::onSeachTextChange)
    }
}
